src/v/kafka/server/quota_manager.h

```diff
@@ -71,6 +71,16 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
         std::optional<atomic_token_bucket> pm_rate;
     };
 
+    // Note: the use of std::shared_ptr<> is generally discouraged in the
+    // redpanda codebase because of the overhead of the atomic reference count
+    // and because seastar's memory model expects data to be deallocated on the
+    // same shard as where it was allocated and sharing std::shared_ptr<>'s
+    // across cores makes it easy to have the deallocation be on a different
+    // shard. The reason why it is acceptable to use a std::shared_ptr<> here is
+    // because (1) we're always allocating and deallocating the objects inside
+    // std::shared_ptr<> on shard 0 and (2) we need a way to keep track of a
+    // reference count across multiple shards, which is exactly what
+    // std::shared_ptr<> is meant for.
```

Comment:
This is a good comment

Reply:
we could create a new shared pointer helper that does this without atomics. handing out a reference via foreign pointer to another core would be +1 to the reference count. then on the foreign core we could allocate a proxy object which would allow that core to further hand out core-local references.

Reply:
thanks for the comment, though. excellent!

Reply:
> we could create a new shared pointer helper that does this without atomics. handing out a reference via foreign pointer to another core would be +1 to the reference count. then on the foreign core we could allocate a proxy object which would allow that core to further hand out core-local references.

I thought about doing something like that, but I went against that because I think that would make `gc()` considerably more expensive. I also don't think the overhead of the `std::shared_ptr<>` copy on the request path when we first see a client id is significant. On GC, he `foreign_ptr<>` would do a cross-core invoke on every deallocation inside `do_local_gc()` which I think is much more expensive than the current coordinated local-then-global gc with only an atomic reference count decrement in `std::~shared_ptr<>` inside `do_local_gc()`. Of course we would need to measure it to know, but the current approach I think would be less expensive and I think is simpler too. (Granted your helper would probably generalize better to other use cases though.)

Reply:
nice write up. 

> more expensive than the current coordinated local-then-global gc with only an atomic reference count decrement in std::~shared_ptr<>

it's tough to say without measuring, as you point out.

when memory is freed on a core that isn't the core on which the memory was allocated then seastar has this atomic compare and swap loop that reaps the frees between the cores. the foreign pointer avoids this by ensuring the free happens on the original core. the presumption here is that the x-core message is ultimately more efficient.

i suspect this is fine, like you point out, but it's not obvious without measuring. on the flip side, if we started adding shared_ptr<T>'s to the system, then we might be going down the wrong path since seastar explicitly tries to avoid shared state.

anyway, fun discussion. i don't think there is anything to do here.